tests/php/ORM/ManyManyThroughListTest.php

```diff
@@ -362,4 +362,52 @@ public function testDefaultSortOnJoinAndMain()
         $this->assertSame('International', $firstReverse->Title);
         $this->assertSame('Argentina', $secondReverse->Title);
     }
+
+    public function testRemoveCallbackOnRemove()
+    {
+        $removedIds = [];
+
+        $base = $this->objFromFixture(ManyManyThroughListTest\TestObject::class, 'parent1');
+        $relation = $base->Items();
+        $remove = $relation->First();
+
+        $relation->setRemoveCallback(function ($list, $ids) use (&$removedIds) {
+            $removedIds = $ids;
+        });
+
+        $relation->remove($remove);
+        $this->assertEquals([$remove->ID], $removedIds);
+    }
+
+    public function testRemoveCallbackOnRemoveById()
+    {
+        $removedIds = [];
+
+        $base = $this->objFromFixture(ManyManyThroughListTest\TestObject::class, 'parent1');
+        $relation = $base->Items();
+        $remove = $relation->First();
+
+        $relation->setRemoveCallback(function ($list, $ids) use (&$removedIds) {
```

Comment:
I would probably structure the test in a manner similar to how the method is intended to be used: by making a dummy object with a custom relation that has setAddCallback and setRemoveCallback defined.

Reply:
Hm not sure I agree, this would reduce the clarity of the test, and really make it an integration test. The fact that you're limited to using these callbacks on overloaded relationship getters is external to the concerns of this unit (`ManyManyThroughList`). And it might change in the future if we figure out a way to implement relationship caching consistently. I'm assuming that we already unit test "relationship getters can be overloaded" elsewhere.

Reply:
OK fair enough.